spring-cloud-kubernetes-client-autoconfig/src/main/java/org/springframework/cloud/kubernetes/client/KubernetesClientAutoConfiguration.java

```diff
@@ -40,8 +42,11 @@
 
 	@Bean
 	@ConditionalOnMissingBean
-	public ApiClient apiClient() {
+	public ApiClient apiClient(KubernetesClientProperties properties) {
 		ApiClient apiClient = kubernetesApiClient();
+		String userAgent = StringUtils.hasText(properties.getUserAgent()) ? properties.getUserAgent()
```

Comment:
Why not just have the default value be set in `KubernetesClientProperties`?

Reply:
I can't, it will mess with `fabric8` order in such a case : 

```
.withUserAgent(or(kubernetesClientProperties.getUserAgent(),
						or(base.getUserAgent(), KubernetesClientProperties.DEFAULT_USER_AGENT)))
				.build();
```

Reply:
What do you mean mess with this order?

Reply:
let's see, may be I miss-understood your comment. I understood this:

> Why not just have the default value be set in KubernetesClientProperties?

as : 

```
	public String getUserAgent() {
		return userAgent;
	}
```

woud become:

```
	public String getUserAgent() {
		return StringUtils.hasLength(userAgent) ? userAgent : "Spring-Cloud-Kubernetes-Application";
	}
```

If this is the case, this would work just fine for `k8s-native` client, but not for `fabric8`, because for `fabric8` it would always ignore `base.getUserAgent()` in : 

```
.withUserAgent(
    or(kubernetesClientProperties.getUserAgent(),
   or(base.getUserAgent(), KubernetesClientProperties.DEFAULT_USER_AGENT)))
.build();
```

because `kubernetesClientProperties.getUserAgent()` would always provide a value...

---

Now that I think about it, it really seems you mean a rather different thing. If so, sorry for not getting

Reply:
Actually I was thinking this in `KubernetesClientProperties`
`private String userAgent = DEFAULT_USER_AGENT`

This would still result in the same problem though.  But I think we should check if `getUserAgent` returns the default value (meaning it was not set explicitly) then check `base.getUserAgent()`

Reply:
I honestly can connect the dots in my head, can you please post some code here? This is the order I wanted to do:

 - `spring.cloud.kubernetes.client.userAgent`
 - base.getUserAgent()
 - `"Spring-Cloud-Kubernetes-Application"`

what is your suggestion here in order to keep the order the same?

Reply:
I think something like this would work.
in `KubernetesClientProperties`
```
private String userAgent = DEFAULT_USER_AGENT
```
Then in `Fabric8AutoConfiguration`
```
String userAgent = or(base.getUserAgent(), KubernetesClientProperties.DEFAULT_USER_AGENT);
if (!KubernetesClientProperties.DEFAULT_USER_AGENT.equals(kubernetesClientProperties.getUserAgent())) {
			userAgent = kubernetesClientProperties.getUserAgent();
}
....
.withUserAgent(userAgent);
```

Reply:
oh wow! :) I would have never even thought about it like that. For me, this:

```
.withUserAgent(or(
    kubernetesClientProperties.getUserAgent(),
    or(base.getUserAgent(), KubernetesClientProperties.DEFAULT_USER_AGENT))
)
.build();
```

is more readable, but I have no strong opinion here.

Reply:
that would not work, needs slight adjustments: 

```
String userAgent = or(base.getUserAgent(), KubernetesClientProperties.DEFAULT_USER_AGENT);
if (StringUtils.hasText(kubernetesClientProperties.getUserAgent())) {
	userAgent = kubernetesClientProperties.getUserAgent();
}
```